service/src/registry.ts

```diff
@@ -556,6 +638,443 @@ export class Registry {
+   * @param {Function} done         The function to call when the operation is
+   *                                complete. `done` will be called with three
+   *                                arguments: an Error object (can be null), the
+   *                                body of the response, and a transport-specific
+   *                                response object useful for logging or
+   *                                debugging.
+   *
+   * @throws {ReferenceError}       If the module or done argument is falsy.
+   * @throws {ArgumentError}        If the module object is missing a deviceId or moduleId value.
+   */
+  addModule(module: Module, done: Registry.ResponseCallback): void {
+    /*Codes_SRS_NODE_IOTHUB_REGISTRY_18_026: [The `addModule` method shall throw a `ReferenceError` exception if `module` or `done` is falsy. ]*/
+    if (!module) throw new ReferenceError('Argument \'module\' cannot be falsy');
+    if (!done) throw new ReferenceError('Argument \'done\' cannot be falsy');
+
+    /*Codes_SRS_NODE_IOTHUB_REGISTRY_18_027: [The `addModule` method shall throw an `ArgumentError` exception if `module.deviceId` or `module.moduleId` is falsy. ]*/
+    if (!module.deviceId) throw new ArgumentError('deviceId property is missing from module object');
+    if (!module.moduleId) throw new ArgumentError('moduleId property is missing from module object');
+
+    /*Codes_SRS_NODE_IOTHUB_REGISTRY_18_028: [The `addModule` method shall construct an HTTP request using information supplied by the caller, as follows:
+    ```
+    PUT /devices/<encodeURIComponent(module.deviceId)/modules/<encodeURIComponent(module.moduleId)>>?api-version=<version> HTTP/1.1
+    Authorization: <sharedAccessSignature>
+    Content-Type: application/json; charset=utf-8
+    Request-Id: <guid>
+
+    <module>
+    ```
+    ]*/
+    const preparedModule = JSON.parse(JSON.stringify(module));
+    this._normalizeAuthentication(preparedModule);
+
+    const path = `/devices/${encodeURIComponent(preparedModule.deviceId)}/modules/${encodeURIComponent(preparedModule.moduleId)}${endpoint.versionQueryString()}`;
+    const httpHeaders = {
+      'Content-Type': 'application/json; charset=utf-8'
+    };
+
+    this._restApiClient.executeApiCall('PUT', path, httpHeaders, preparedModule, done);
+  }
+
+  /**
+   * @method            module:azure-iothub.Registry#getModulesOnDevice
+   * @description       Get a list of all modules on an IoT Hub device
+   *
+   * @param {String}    deviceId  ID of the device we're getting modules for
+   * @param {Function}  done      The callback which will be called with either an Error object
+   *                              or an array of module:azure-iothub.Module objects
+   *                              for all the modules.
+   *
+   * @throws {ReferenceError}     If the deviceId or done argument is falsy.
+   */
+  getModulesOnDevice(deviceId: string, done: (err: Error, modules?: Module[], response?: any) => void): void {
```

Comment:
>Module[] [](start = 68, length = 8)

page-able query interface instead?

Reply:
Investigating.  I know there's a limit to 10 modules for Edge.  I'm not sure about the general case.

---
In reply to: [176870694](https://github.com/Azure/azure-iot-sdk-node/pull/256#discussion_r176870694) [](ancestors = 176870694)